src/python/pants/engine/fs.py

```diff
@@ -76,33 +76,15 @@ def __init__(
                       with `!`, e.g. `!ignore.py`.
         :param glob_match_error_behavior: whether to warn or error upon match failures
         :param conjunction: whether all `include`s must match or only at least one must match
-        :param description_of_origin: a human-friendly description of where this PathGlobs request is
-                                      coming from, used to improve the error message for unmatched
-                                      globs. For example, this might be
-                                      "the option `--isort-config`".
+        :param description_of_origin: a human-friendly description of where this PathGlobs request
+                                      is coming from, used to improve the error message for
+                                      unmatched globs. For example, this might be the option
+                                      `--isort-config`.
         """
         self.globs = tuple(globs)
         self.glob_match_error_behavior = glob_match_error_behavior
         self.conjunction = conjunction
         self.description_of_origin = description_of_origin or ""
-        self.__post_init__()
-
-    def __post_init__(self) -> None:
-        if self.glob_match_error_behavior == GlobMatchErrorBehavior.ignore:
-            if self.description_of_origin:
-                raise ValueError(
-                    "You provided a `description_of_origin` value when `glob_match_error_behavior` is set to "
-                    "`ignore`. The `ignore` value means that the engine will never generate an error when "
-                    "the globs are generated, so `description_of_origin` won't end up ever being used. "
-                    "Please either change `glob_match_error_behavior` to `warn` or `error`, or remove "
-                    "`description_of_origin`."
-                )
-        else:
-            if not self.description_of_origin:
-                raise ValueError(
-                    "Please provide a `description_of_origin` so that the error message is more helpful to "
-                    "users when their globs fail to match."
-                )
```

Comment:
Why are you removing this? It's meant to ensure that rule authors provide `description_of_origin` when we might error. 

When we can't error, this validation is meant to avoid the user getting a false sense of security - we want them to be well aware that it's impossible to error.

Reply:
Ah, I see. `./pants --glob-expansion-failure=ignore filedeps2 src/python/pants/util:collections` will fail currently. That's not great..my bad for not catching that use case - generally, the `GlobMatchErrorBehavior` is not computed at runtime like this.

Rather than completely removing `__post_init__()`, can you please only remove the first part `if self.glob_match_error_behavior == GlobMatchErrorBehavior.ignore:`? We should still error if `GlobMatchErrorBehavior` is `warn` or `error` and the rule author did not provide a `description_of_origin`.

Reply:
Conditionally checking the `self.description_of_origin` seemed really confusing and weird to me, and would have required fixing every callsite to avoid passing in `description_of_origin` iff `--files-not-found-behavior=ignore`. I'll try to figure out how to preserve the intent here while not breaking the `ignore` case.

Reply:
Iirc there are two call sites, I think both in graph.py. One for hydrating the sources field and another for hydrating the bundles field. 

That would be awesome if you could dynamically pass in `description_of_origin` based on the field. I know the call sites will look really hacky and ugly, but generally we don’t calculate this value dynamically and it would be nice to keep the original validation so that we nudge rule authors to warn or error when it seems that was the their intent. 

If that doesn’t work, I’m okay with removing that first check for when the value is ignore. I would like to keep the second check for when they say to warn/error but don’t give a description, though. (And we need it to avoid a cryptic rust runtime error).

Reply:
> Iirc there are two call sites

There are many `PathGlobs()` constructions throughout the codebase -- feel like I'm missing your meaning here.

> That would be awesome if you could dynamically pass in `description_of_origin` based on the field.

Could you elaborate on this? I'm not sure I understand.

> it would be nice to keep the original validation so that we nudge rule authors to warn or error when it seems that was the their intent

This seems really reasonable. Not sure how to do it yet.

Reply:
Sorry for the prior brevity, I was on my phone.

> There are many PathGlobs() constructions throughout the codebase -- feel like I'm missing your meaning here.

What I meant is that I believe there are only two call sites where the behavior is determined by `FilesNotFoundBehavior`. Indeed, there are many `PathGlobs` usages, but the majority of them hardcode what the `GlobMatchErrorBehavior` should be, e.g.:

https://github.com/pantsbuild/pants/blob/63e70c17732058393f13221ecbefeae74125ddcb/src/python/pants/backend/python/lint/isort/rules.py#L94-L99

Precisely, anywhere that we request the `GlobMatchErrorBehavior` in a rule signature, we must add this workaround. I believe this is only the rules `hydrate_sources` and `hydrate_bundles`. (We used to use the singleton to determine the error behavior for unmatched file args, but always error now.)

> Could you elaborate on this? I'm not sure I understand.

Here's a diff that fixes both cases.

```diff
diff --git a/src/python/pants/engine/legacy/graph.py b/src/python/pants/engine/legacy/graph.py
index 5cab62bfc..79cbf3b62 100644
--- a/src/python/pants/engine/legacy/graph.py
+++ b/src/python/pants/engine/legacy/graph.py
@@ -717,7 +717,11 @@ async def hydrate_sources(
         glob_match_error_behavior=glob_match_error_behavior,
         # TODO(#9012): add line number referring to the sources field. When doing this, we'll likely
         # need to `await Get[BuildFileAddress](Address)`.
-        description_of_origin=f"{address}'s `{sources_field.arg}` field",
+        description_of_origin=(
+            f"{address}'s `{sources_field.arg}` field"
+            if glob_match_error_behavior != GlobMatchErrorBehavior.ignore
+            else None
+        ),
     )
     snapshot = await Get[Snapshot](PathGlobs, path_globs)
     fileset_with_spec = _eager_fileset_with_spec(
@@ -740,7 +744,11 @@ async def hydrate_bundles(
             glob_match_error_behavior=glob_match_error_behavior,
             # TODO(#9012): add line number referring to the bundles field. When doing this, we'll likely
             # need to `await Get[BuildFileAddress](Address)`.
-            description_of_origin=f"{address}'s `bundles` field",
+            description_of_origin=(
+                f"{address}'s `bundles` field"
+                if glob_match_error_behavior != GlobMatchErrorBehavior.ignore
+                else None
+            ),
         )
         for pg in bundles_field.path_globs_list
     ]
```

Reply:
Done!!! Thank you for explaining this!!!